airflow/api/common/experimental/mark_tasks.py

```diff
@@ -77,12 +88,14 @@ def set_state(task, execution_date, upstream=False, downstream=False,
     """
     assert timezone.is_localized(execution_date)
 
-    assert task.dag is not None
-    dag = task.dag
+    task_dags = {task.dag for task in tasks}
+    assert len(task_dags) == 1
```

Comment:
`assert` will be disabled at runtime if python is run with omptimizations, so it _may_ be a bad idea to check this invariant like this. (i.e. maybe we should replace it with a `raise Something()`

Reply:
Good catch. There are a few `assert` statements in this file--I'll change them all to raise exceptions.